File: YuGiOh2/Cards/C97169186.cs
using System.Linq;
using YuGiOh2.Base;

namespace YuGiOh2.Cards
{
    /// <summary>
    /// 地碎
    /// </summary>
    public class C97169186
    {
        public static int Type { get; } = (int)ChooseTargetType.None;

        public static bool CheckIfAvailable(Player player)
        {
            return player.Enemy.Field.MonsterFields.Any(c => c != null && !c.Status.FaceDown);
        }

        public static void ProcessEffect(Player player)
        {
            var max = player.Enemy.Field.MonsterFields.Select(c => c == null ? 0 : c.DEF).Max();
            for (int i = 0; i < 5; i++)
            {
                if (player.Enemy.Field.MonsterFields[i] == null || 
                    player.Enemy.Field.MonsterFields[i].Status.FaceDown || 
                    player.Enemy.Field.MonsterFields[i].DEF != max)
                    continue;

                player.Enemy.AddCardToGrave(player.Enemy.Field.MonsterFields[i]);
                player.Enemy.Field.MonsterFields[i] = null;
                return;
            }
        }
    }
}

File: YuGiOh2/Scripts/C97169186.lua
Type = 0;
--地碎
function checkIfAvailable(player)
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and not monster.Status.FaceDown then
            return true;
        end
    end
    return false;
end

function processEffect(player)
    if targetID == nil then
        return ;
    end
    local max = 0;
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and monster.ATK > max then
            max = monster.ATK;
        end
    end
    for i, monster in pairs(player.Enemy.Field.MonsterFields) do
        if monster ~= nil and monster.ATK == max then
            player.Enemy:AddCardToGrave(monster);
            player.Enemy:ClearMonsterField(i - 1);
        end
    end
end